arangod/Replication/DatabaseInitialSyncer.cpp

```diff
@@ -1009,111 +1014,152 @@ Result DatabaseInitialSyncer::fetchCollectionSyncByKeys(arangodb::LogicalCollect
-                    std::string(
-                        "timed out waiting for response from leader at ") +
-                        _config.leader.endpoint);
+      return replutils::buildHttpError(response.get(), url, _config.connection);
     }
 
-    if (isAborted()) {
-      stats.waitedForInitial += waitTime;
-      return Result(TRI_ERROR_REPLICATION_APPLIER_STOPPED);
+    VPackBuilder builder;
+    Result r = replutils::parseResponse(builder, response.get());
+
+    if (r.fail()) {
+      ++stats.numFailedConnects;
+      return Result(TRI_ERROR_REPLICATION_INVALID_RESPONSE,
+                    concatT("got invalid response from leader at ",
+                            _config.leader.endpoint, url, ": ", r.errorMessage()));
     }
 
-    std::chrono::milliseconds sleepTime = ::sleepTimeFromWaitTime(waitTime);
-    std::this_thread::sleep_for(sleepTime);
-  }
-  
-  stats.waitedForInitial += TRI_microtime() - startTime;
+    VPackSlice const slice = builder.slice();
+    if (!slice.isObject()) {
+      ++stats.numFailedConnects;
+      return Result(TRI_ERROR_REPLICATION_INVALID_RESPONSE,
+                    std::string("got invalid response from leader at ") +
+                    _config.leader.endpoint + url + ": response is no object");
+    }
 
-  if (replutils::hasFailed(response.get())) {
-    ++stats.numFailedConnects;
-    return replutils::buildHttpError(response.get(), url, _config.connection);
-  }
+    return Result();
 
-  VPackBuilder builder;
-  Result r = replutils::parseResponse(builder, response.get());
+  };
 
-  if (r.fail()) {
-    ++stats.numFailedConnects;
+  auto ck = keysCall(true);
+  if (!ck.ok()) {
+    return ck;
+  }
+  VPackSlice const c = slice.get("count");
+  uint64_t ndocs = 0;
+  if (c.isNumber()) {
+    ndocs = c.getNumber<uint64_t>();
+  } else {
     return Result(TRI_ERROR_REPLICATION_INVALID_RESPONSE,
-                  concatT("got invalid response from leader at ",
-                          _config.leader.endpoint, url, ": ", r.errorMessage()));
+                  std::string("got invalid response from master at ") +
+                  _config.leader.endpoint + url + ": response count not a number");
   }
 
-  VPackSlice const slice = builder.slice();
-  if (!slice.isObject()) {
-    ++stats.numFailedConnects;
-    return Result(TRI_ERROR_REPLICATION_INVALID_RESPONSE,
-                  std::string("got invalid response from leader at ") +
-                      _config.leader.endpoint + url + ": response is no object");
+#ifdef ARANGODB_ENABLE_FAILURE_TESTS
+  if (ndocs > _quickKeysNumDocsLimit && slice.hasKey("id")) {
+    LOG_TOPIC("6e1b3", ERR, Logger::REPLICATION)
+        << "client: DatabaseInitialSyncer::run - expected only document count for quick call";
+    TRI_ASSERT(false);
+  }
+#endif
+
+  if (!slice.hasKey("id")) { // we only have count
+    if (ndocs >= 1000000) { // only bother beyond a million documents
+      maxWaitTime = ndocs * 8 / 100000;
```

Comment:
maxWaitTime as used above (https://github.com/arangodb/arangodb/pull/13394/files?diff=unified&w=1#diff-db234bdc91c0f395720e4bf5d69da8fbc541f5d66a6f73aac78ab6df02b7a387R1086) has a unit of microseconds. 
If this formula produces a result in microseconds, then for 1M documents the maxWaitTime is 80 microseconds, which is probably wait too low.
What we should do here is to do a calculation of the wait time (in the correct unit) and then make sure whatever we calculated does not go below the configured wait time (i.e. _config.applier._initialSyncMaxWaitTime). Otherwise it is not an improvement, but a pessimization.